### PR TITLE
Pojo Updates and improvements.

### PR DESCRIPTION
**Pull request by rozza**
First 4 Commits are community based (with minor tweaks).

* 37ccb33 - Improves the functionality and usability `@BsonCreator` and `@BsonProperty` annotations.
* 50db877 - Allows discriminators to be used in a Collection.
* 2714656 - Fixes a range of issues regarding containers, by introducing a new `PropertyCodecProvider` abstraction. Biggest change here with a new user facing API. The abstraction is nice in that it also lays the groundwork for supporting custom maps.
* 22b42f8 - Improves the support for sub classes with the codec.

Extra commits are:

* 972216e - A simple test to prove a bug was fixed.
* dc9565f - Removing the validation in `TypeData`. The new  `PropertyCodecProvider`  abstraction means that users should have finer grained control over the concept of what are valid Maps / Collections.

### CODE REVIEW

File: bson/src/main/org/bson/codecs/pojo/CollectionPropertyCodecProvider.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecConfigurationException;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+
+import static java.lang.String.format;
+
+final class CollectionPropertyCodecProvider implements PropertyCodecProvider {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/CollectionPropertyCodecProvider.java]: This now handles Collections, custom collections can now be supported by user supplied `PropertyCodecProvider` implementations.

File: bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java [hunk 2]
```diff
@@ -111,49 +112,104 @@ private void processPropertyAnnotations(final ClassModelBuilder<?> classModelBui
...
-                    if (annotation.annotationType().equals(BsonCreator.class)) {
-                        if (creatorExecutable != null) {
-                            throw new CodecConfigurationException("Found multiple constructors / methods annotated with @BsonCreator");
-                        } else if (!clazz.isAssignableFrom(method.getReturnType())) {
-                            throw new CodecConfigurationException(
-                                    format("Invalid method annotated with @BsonCreator. Returns '%s', expected %s", method.getReturnType(),
-                                            clazz));
+        Class<?> bsonCreatorClass = clazz;
+        boolean foundStaticBsonCreatorMethod = false;
+        while (bsonCreatorClass != null && !foundStaticBsonCreatorMethod) {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java]: Ensures `@BsonCreator` can be defined by super classes.

File: bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java [hunk 3]
```diff
@@ -111,49 +112,104 @@ private void processPropertyAnnotations(final ClassModelBuilder<?> classModelBui
...
+                boolean isIdProperty = creatorExecutable.getIdPropertyIndex() != null && creatorExecutable.getIdPropertyIndex().equals(i);
                 Class<?> parameterType = parameterTypes.get(i);
-                PropertyModelBuilder<?> propertyModelBuilder = classModelBuilder.getProperty(bsonProperty.value());
-                if (propertyModelBuilder == null) {
-                    addCreatorPropertyToClassModelBuilder(classModelBuilder, bsonProperty.value(), parameterType);
-                } else if (!propertyModelBuilder.getTypeData().isAssignableFrom(parameterType)) {
+                Type genericType = parameterGenericTypes.get(i);
+                PropertyModelBuilder<?> propertyModelBuilder = null;
+
+                if (isIdProperty) {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java]: Th `@bsonId` annotation is now supported with `@BsonCreator`.

File: bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java [hunk 4]
```diff
@@ -111,49 +112,104 @@ private void processPropertyAnnotations(final ClassModelBuilder<?> classModelBui
...
+                    propertyModelBuilder = classModelBuilder.getProperty(classModelBuilder.getIdPropertyName());
+                } else {
+                    BsonProperty bsonProperty = properties.get(i);
+
+                    // Find the property using write name and falls back to read name
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java]: We now prefer the write name first for properties. This removes the confusion about which of the 3 property names should be used with `@BsonCreator`.  Legacy property names are still supported if needs be.

File: bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java [hunk 5]
```diff
@@ -111,49 +112,104 @@ private void processPropertyAnnotations(final ClassModelBuilder<?> classModelBui
...
 
-    private <T, S> void addCreatorPropertyToClassModelBuilder(final ClassModelBuilder<T> classModelBuilder, final String name,
-                                                              final Class<S> clazz) {
-        classModelBuilder.addProperty(createPropertyModelBuilder(new PropertyMetadata<S>(name, classModelBuilder.getType().getSimpleName(),
-                TypeData.builder(clazz).build())).readName(null).writeName(name));
+    @SuppressWarnings("unchecked")
+    private static <T> void tryToExpandToGenericType(final Class<?> parameterType, final PropertyModelBuilder<T> propertyModelBuilder,
+                                                     final Type genericType) {
+        if (parameterType.isAssignableFrom(propertyModelBuilder.getTypeData().getType())) {
+            // The existing getter for this field returns a more specific type than what the constructor accepts
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java]: This allows covariant types to be supported by the `@BsonCreator`.

File: bson/src/main/org/bson/codecs/pojo/InvalidMapImplementation.java [hunk 6]
```diff
@@ -1,84 +0,0 @@
...
- * limitations under the License.
- */
-
-package org.bson.codecs.pojo;
-
-import java.util.Collection;
-import java.util.Map;
-import java.util.Set;
-
-final class InvalidMapImplementation implements Map<Map<String, Integer>, Integer> {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/InvalidMapImplementation.java]: This is no longer used and shouldn't have been created in src/main.

File: bson/src/main/org/bson/codecs/pojo/MapPropertyCodecProvider.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecConfigurationException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import static java.lang.String.format;
+
+final class MapPropertyCodecProvider implements PropertyCodecProvider {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/MapPropertyCodecProvider.java]: Like the `CollectionPropertyCodecProvider` this replaces the old Pojo Map Codec and handles validation.

File: bson/src/main/org/bson/codecs/pojo/MapPropertyCodecProvider.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public <T> Codec<T> get(final TypeWithTypeParameters<T> type, final PropertyCodecRegistry registry) {
+        if (Map.class.isAssignableFrom(type.getType()) && type.getTypeParameters().size() == 2) {
+            Class<?> keyType = type.getTypeParameters().get(0).getType();
+            if (!keyType.equals(String.class)) {
+                throw new CodecConfigurationException(format("Invalid Map type. Maps MUST have string keys, found %s instead.", keyType));
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/MapPropertyCodecProvider.java]: Previously, TypeData would have thrown an IllegalArgument Exception for a Map without a String key.

File: bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java [hunk 9]
```diff
@@ -184,41 +197,10 @@ private void decodeProperties(final BsonReader reader, final DecoderContext deco
 
     private <S> void addToCache(final PropertyModel<S> propertyModel) {
         Codec<S> codec = propertyModel.getCodec() != null ? propertyModel.getCodec()
-                : specializePojoCodec(propertyModel, getCodecFromTypeData(propertyModel.getTypeData()));
+                : specializePojoCodec(propertyModel, propertyCodecRegistry.get(propertyModel.getTypeData()));
         propertyModel.cachedCodec(codec);
     }
 
-    @SuppressWarnings({"rawtypes", "unchecked"})
-    private  <S> Codec<S> getCodecFromTypeData(final TypeData<S> typeData) {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java]: The new PropertyCodecProviders replace these all pojo based codecs.

File: bson/src/main/org/bson/codecs/pojo/CollectionPropertyCodecProvider.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public <T> Codec<T> get(final TypeWithTypeParameters<T> type, final PropertyCodecRegistry registry) {
+        if (Collection.class.isAssignableFrom(type.getType())) {
+            if (type.getTypeParameters().size() == 0) {
+                throw new CodecConfigurationException("Invalid Collection type. Collections must have a defined type.");
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/CollectionPropertyCodecProvider.java]: Previously, TypeData would have thrown an IllegalArgument Exception for a Collection without a defined type

File: bson/src/main/org/bson/codecs/pojo/PropertyReflectionUtils.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+package org.bson.codecs.pojo;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static java.lang.reflect.Modifier.isPublic;
+
+final class PropertyReflectionUtils {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PropertyReflectionUtils.java]: This was moved / abstracted out of `PojoBuilderHelper`.

File: bson/src/main/org/bson/codecs/pojo/TypeWithTypeParameters.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.List;
+
+/**
+ * A combination of a type and its type parameters.
+ *
+ * @param <T> the type which potentially has parameterized types
+ * @since 3.6
+ */
+public interface TypeWithTypeParameters<T> {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/TypeWithTypeParameters.java]: Extended by `TypeData` and used by `PropertyCodecProviders` without making `TypeData` public.

File: bson/src/main/org/bson/codecs/pojo/annotations/BsonId.java [hunk 13]
```diff
@@ -32,6 +32,6 @@
  */
 @Documented
 @Retention(RetentionPolicy.RUNTIME)
-@Target({ElementType.FIELD, ElementType.METHOD})
+@Target({ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER})
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/annotations/BsonId.java]: `@BsonCreator` now supports `@BsonId`

File: bson/src/test/unit/org/bson/codecs/pojo/ClassModelBuilderTest.java [hunk 14]
```diff
@@ -107,10 +106,10 @@ public void testFieldsMappedClassTypes() {
         ClassModelBuilder<ConcreteCollectionsModel> builder =
                 ClassModel.builder(ConcreteCollectionsModel.class);
 
-        assertEquals(ArrayList.class, builder.getProperty("collection").getTypeData().getType());
-        assertEquals(ArrayList.class, builder.getProperty("list").getTypeData().getType());
+        assertEquals(Collection.class, builder.getProperty("collection").getTypeData().getType());
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/ClassModelBuilderTest.java]: We no longer need to alter the type data for interfaces to concrete implementations.

File: bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java [hunk 15]
```diff
@@ -262,6 +270,47 @@ public void testCanHandleTopLevelGenericIfHasCodec() {
                 "{'myGenericField': {'$numberLong': '5'}}");
     }
 
+    @Test
+    public void testCustomRegisteredPropertyCodecWithValue() {
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java]: Support for custom containers like `Optional` is now easy with a custom `PropertyCodecProvider`.

File: bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java [hunk 16]
```diff
@@ -262,6 +270,47 @@ public void testCanHandleTopLevelGenericIfHasCodec() {
...
+    }
+
+    @Test(expected = CodecConfigurationException.class)
+    public void testDecodingInvalidCollectionModel() {
+        decodingShouldFail(getCodec(InvalidCollectionModel.class), "{collectionField: [1, 2, 3]}");
+    }
+
+    @Test
+    public void testInvalidMapModelWithCustomPropertyCodecProvider() {
+        encodesTo(getPojoCodecProviderBuilder(InvalidMapModel.class).register(new InvalidMapPropertyCodecProvider()), getInvalidMapModel(),
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/PojoCustomTest.java]: Custom `PropertyCodecProviders` also allow us to handle previously invalid Map implementations.

File: bson/src/test/unit/org/bson/codecs/pojo/PojoRoundTripTest.java [hunk 17]
```diff
@@ -105,10 +121,19 @@ public void test() {
                         + "'child': {'_id': 'child', 'myFinalField': 10, 'myIntField': 10,"
                         + "'model': {'integerField': 42, 'stringField': 'myString'}}}"));
 
+        data.add(new TestData("BsonIgnore invalid map", new BsonIgnoreInvalidMapModel("myString"),
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/PojoRoundTripTest.java]: As TypeData no longer validates types, the conventions can now ignore fields that previously would have thrown a validation error before conventions could be applied.

File: bson/src/test/unit/org/bson/codecs/pojo/TypeDataTest.java [hunk 18]
```diff
@@ -79,37 +77,4 @@ public void testRecursiveTypeData() {
 
         typeData.toString();
     }
-
-    @Test(expected = IllegalStateException.class)
-    public void testListNoParamsValidation() {
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/TypeDataTest.java]: TypeData no longer validates the types, allowing users to provide custom ways to support the types, or the default  `PropertyCodecProviders` will validate them.

File: bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java [hunk 19]
```diff
@@ -44,21 +44,29 @@
...
-        this(classModel, registry, discriminatorLookup, new ConcurrentHashMap<ClassModel<?>, Codec<?>>(), shouldSpecialize(classModel));
+    PojoCodecImpl(final ClassModel<T> classModel, final CodecRegistry registry, final List<PropertyCodecProvider> propertyCodecProviders,
+                  final DiscriminatorLookup discriminatorLookup) {
+        this(classModel, registry, propertyCodecProviders, null, discriminatorLookup,
+                new ConcurrentHashMap<ClassModel<?>, Codec<?>>(), shouldSpecialize(classModel));
     }
 
-    PojoCodecImpl(final ClassModel<T> classModel, final CodecRegistry registry, final DiscriminatorLookup discriminatorLookup,
+    PojoCodecImpl(final ClassModel<T> classModel, final CodecRegistry registry, final List<PropertyCodecProvider> propertyCodecProviders,
+                  final PropertyCodecRegistry propertyCodecRegistry, final DiscriminatorLookup discriminatorLookup,
```
Comment (jyemin) [inline bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java]: These two constructors are confusing as in this one a non-null propertyCodecProviders is ignored if propertyCodecRegistry is also non-null.

Instead have one constructor that takes just a List<PropertyCodecProvider> and another takes just a PropertyCodecRegistry

File: bson/src/test/unit/org/bson/codecs/pojo/entities/OptionalPropertyCodecProviderForTest.java [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+import org.bson.BsonReader;
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.pojo.PropertyCodecProvider;
+import org.bson.codecs.pojo.PropertyCodecRegistry;
+import org.bson.codecs.pojo.TypeWithTypeParameters;
+
+public class OptionalPropertyCodecProviderForTest implements PropertyCodecProvider {
```
Comment (jyemin) [inline bson/src/test/unit/org/bson/codecs/pojo/entities/OptionalPropertyCodecProviderForTest.java]: Rename to OptionalPropertyCodecProvider

File: bson/src/main/org/bson/codecs/pojo/InstanceCreatorImpl.java [hunk 21]
```diff
@@ -41,9 +41,18 @@
...
             this.cachedValues = new HashMap<PropertyModel<?>, Object>();
             this.properties = new HashMap<String, Integer>();
+
+            if (creatorExecutable.getIdPropertyIndex() != null) {
+              this.properties.put(ClassModelBuilder.ID_PROPERTY_NAME, creatorExecutable.getIdPropertyIndex());
+            }
+
             for (int i = 0; i < creatorExecutable.getProperties().size(); i++) {
-                this.properties.put(creatorExecutable.getProperties().get(i).value(), i);
+                if (creatorExecutable.getIdPropertyIndex() == null || !creatorExecutable.getIdPropertyIndex().equals(i)) {
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson/src/main/org/bson/codecs/pojo/InstanceCreatorImpl.java]: Done

File: bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java (hunk 19 above)
Comment (rozza) [reply bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java]: Done

File: bson/src/test/unit/org/bson/codecs/pojo/entities/OptionalPropertyCodecProviderForTest.java (hunk 20 above)
Comment (rozza) [reply bson/src/test/unit/org/bson/codecs/pojo/entities/OptionalPropertyCodecProviderForTest.java]: Done

File: bson/src/test/unit/org/bson/codecs/pojo/entities/conventions/SuperClassModel.java [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.pojo.entities.conventions;
+
+import org.bson.codecs.pojo.annotations.BsonDiscriminator;
+
+@BsonDiscriminator
+public abstract class SuperClassModel {
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson/src/test/unit/org/bson/codecs/pojo/entities/conventions/SuperClassModel.java]: Turns out we weren't getting annotations from interfaces, so the comment in the commit was void. Works now though :)

File: bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java [hunk 23]
```diff
@@ -49,26 +49,31 @@
     private final ConcurrentMap<ClassModel<?>, Codec<?>> codecCache;
     private final boolean specialized;
 
-    PojoCodecImpl(final ClassModel<T> classModel, final CodecRegistry registry, final List<PropertyCodecProvider> propertyCodecProviders,
-                  final DiscriminatorLookup discriminatorLookup) {
-        this(classModel, registry, propertyCodecProviders, null, discriminatorLookup,
-                new ConcurrentHashMap<ClassModel<?>, Codec<?>>(), shouldSpecialize(classModel));
+    PojoCodecImpl(final ClassModel<T> classModel, final CodecRegistry codecRegistry,
+                  final List<PropertyCodecProvider> propertyCodecProviders, final DiscriminatorLookup discriminatorLookup) {
+        this.classModel = classModel;
```
Comment (jyemin) [inline bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java]: Nit: This constructor should still just call the other.

Comment (jyemin) [reply bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java]: Scratch that.  Thanks for the off-line explanation

**Pull request closed without merging**